### ISSUE TITLE
Update Conflict while editing files on Google Drive (drive G:)

### ISSUE THREAD

**Issue by emanueol** [dbeaver/dbeaver#16952]
#### System information: 
- Windows 10
- DBeaver version 22.1.1.202206261508, release date June 27, 2022
- Java version java version "1.8.0_331", Java(TM) SE Runtime Environment (build 1.8.0_331-b09), Java HotSpot(TM) 64-Bit Server VM (build 25.331-b09, mixed mode)
- Additional extensions Terminal > Terminal

#### Connection specification:
- Database name and version Snowflake
- Driver name net.snowflake:snowflake-jdbc:RELEASE [3.13.6]
- Do you use tunnels or proxies (SSH, SOCKS, etc)? no

#### Describe the problem you're observing:
I have Google Drive installed which creates a drive G: which maps virtually to my cloud space in google.

* While editing --> most of the times I press CTRL+S to save  file --> i see annoying unexpected popup
* 
![image](https://user-images.githubusercontent.com/8422412/177013314-42830d39-2217-4c29-ba61-9d43fba63aa0.png)

No one changed anything outside DBeaver, so I dont understand why every time I save file I see this damm popup suggesting file was changed on file system outside DBeaver and if I want to smag changes I did while editing file in DBEaver ?!?!?
Is this a problem with Google Drive vs DBEaver can only ?

A second different popup also keeps appearing from nowhere, tyhpically when i alt table back to DBEaver:
![image](https://user-images.githubusercontent.com/8422412/177013469-c2c39ada-19df-450b-87a8-26a1ffce87eb.png)

This is really annoying these 2 popups from bellow 2 actions:
- I only edit and save file in DBeaver sql editor.
- and either save a file or alt tab back to DBeaver.

Any suggestions ?

#### Steps to reproduce, if exist:
Simply save a file and popup appears:
![image](https://user-images.githubusercontent.com/8422412/177013398-77169918-6765-4f25-a323-1c5ca33c0848.png)

**Comment by emironovDB** [comment]
@emanueol 
Thanks for the bug report.

**Comment by chrillemus** [comment]
I have the same issue

**Comment by SSzretter** [comment]
I have the same issue in the latest as well.

**Comment by ayocumnd** [comment]
Just need an option to turn off the check. Anything, this killing me.

**Comment by pradernd** [comment]
This issue is still occurring in 23.0.0

**Comment by emanueol** [comment]
Adding some insight that may help devs.
I also had problem with source in google drive but while developing in a c++ context and i figured workaround which you may pursue to understand.

When compiling c++ code 2nd time after running my app and exiting it (my exe not in memory) but somehow deeply in silicon or google drive or some OS pipe i didn't bothered going so deeply to discovered what was still thinking my exe was still running in memory, so compiler couldn't overwrite exe with new compilation.
Seems bit different use case right "file handle locked somehow" vs "DBeaver popups informing external activity on my source files when nothing was edit".. seems in both cases common factor is "google drive".

So i moved my fun side c++ opengl project into dropbox and voila no more problems.

Now i wonder if this can be fixed and/or workarounded by DBeaver devs despite probably problem is with google drive (at least their windows client).

Thats it. Hope this helps you guys.

**Comment by Keper** [comment]
I have the same issue on ubuntu 22.10 and saving on Samba NAS, if I save the files in local there are no problems.
I suppose it depends by a delay to read / write files on the network?

**Comment by SSzretter** [comment]
Just updating that I still have this issue on 23.1.0.2023061220813 - both in the ultimate version and community,   I have noticed it it is not as bad on a different machine but it does happen across all my machines that have google drive.   It would be nice if we could either adjust the file detection time or disable it completely.   Maybe a delay would help or figuring out what google drive is doing to the file that is causing it to look like it has changed.

**Comment by emanueol** [comment]
Just to add that this doesnt happen with Dropbox.
I did some side-fun-development with OpenGL 3D library in C++ where i was having similar problem (c++ compiler couldn't overwrite the .exe upon each new compilation, was hard to understand but then i figured out that moving the .exe c++ source files from GDrive into Dropbox, that problem stopped.

Hope this helps, as i don't know if its a Google Drive problem that Dropbox is better, or if its a client problem like DBeaver (to compile C++ i was using vs code and some command line C++ compiler/debug tools, so i guess the problem being .exe creation, doesn't seem to be a simple client problem on this case. 

But with DBeaver i wonder if:
- you guys can work with GDrive dev team for a workaround ?
- understand why problem doesn't happen with Dropbox, and try replicate workaround for Google Drive ?

Thanks

**Comment by MattAnctil** [comment]
Issue is still present for Win 10 Pro 22H2 and DBeaver 23.2.4.

Could we get at an option in the settings to disable until this is fixed?

**Comment by y2jhotaru** [comment]
The bug still persists in W11Pro with version 24.1.1.202406231636

**Comment by delta4957** [comment]
The same issue is still present in DBeaver Lite Edition, Version 24.1.0.202406100905, Release date: June I0, 2024

**Comment by ShadelessFox** [comment]
I believe the issue is caused by Google Drive preserving file modification timestamps - it seems that it updates them incorrectly, causing Eclipse to think the file was changed. I doubt there's a way to fix it in DBeaver.

The only thing we can do is to ignore such changes via an option in the preferences and either always ignore or apply external changes.

**Comment by emanueol** [comment]
Dbeaver team could (if haven't yet) request fix directly to Google drive
devs, meanwhile im sure dbeaver can envise some workaround.

Best Regards,
Emanuel Oliveira

**Comment by ShadelessFox** [comment]
No, we haven't contacted the Google Drive development team. I doubt they would change anything, as this behavior seems to have existed for the past decade or so.

**Comment by emanueol** [comment]
> No, we haven't contacted the Google Drive development team. I doubt they would change anything, as this behavior seems to have existed for the past decade or so.

Ok good to know.

Meanwhile, I no longer have have this problem for some months.. not sure what changed tho.
I wonder if this still happening with others ? If not we can quietly closed it I guess. Thanks.

**Comment by SSzretter** [comment]
Just want to mention that I have not seen this issue for a long time -
since I purchased and switched to DBeaver Ultimate.   This week I happened
to install Community after wiping my profile and I am receiving the error
again using that version.

**Comment by emanueol** [comment]
I'm starting to have this issue again... every time I press CTL+S to save I get the damm popup saying the file changed on the file system... really annoying. 
Do you have any update on how to fix or work around this?

TLDR: I'm saving files on my Google Drive in Streaming mode.

Here animated gif showing the problem:
- Notice that the timestamps to the second have not changed, despite getting the DBeaver "damm" popup.
- also Process Explorer doesn't show any process with a file handle open on my file.

https://drive.google.com/file/d/1gtqgQFXYjsnULhF2kVGYSGGW78sy_Y2K/view?usp=sharing

And here screenshots showing file timestamp not changed (up to second), but damm popup decided to appear (1st after pressing CTRL+S, and 2nd time without pressing any key, simply by bringing DBeaver to foreground):

<img width="1094" alt="anim gif screenshots of file timestamps" src="https://github.com/user-attachments/assets/b36f73f1-9840-45d3-9018-b37335a94491">

**Comment by emanueol** [comment]
DBeaver team, lets plan a debug session , I fully volunteer for screen-sharing giving you control.

Let me know in private message so I share my personal email.
Im sure we can workaround this, i got some ideas. Thanks.

**Comment by emanueol** [comment]
**UPDATE: Found the problem, as able to replicate the problem after a suggestion of a friend, seems to be with java library EFS (a file system abstraction library/system) and Google Drive.

1. downloaded and installed https://www.eclipse.org/downloads
2. created a .sql file on my Google Drive G:\
3. open .sql on Eclipse > edit > when pressing CTRL+S I see same "damm" popup as DBeaver

The "damm" popups are the same (the ones telling file was changed outside DBeaver or Eclipse, when nothing changed them.
Even the new popup of new DBeaver version telling about to exclude in windows Defenser or not the exe scan, also appears in Eclipse, so cleary this funtionability of EFS abrastracting file system is bugged with Google Drive.

Hope this helps, now that we know exactly where/what is the problem.
![eclipse_Update Conflict while editing files on Google Drive](https://github.com/user-attachments/assets/db46e81f-caab-4119-842d-f275cad61d46)

~~~~~~~~~~~~~~~~~
Meanwhile I found in https://stackoverflow.com/questions/815707/avoiding-eclipse-changed-on-file-system-dialog similar problem with eclipse, mentioning some workarounds:
Eclipse > Preferences->General->Workspace->"Refresh Using Native Hooks or Polling" *OR* "Refresh on Access"

I found these 2 options exists in DBeaver, but problem still persists.
~~~~~~~~~~~~~~~~~
So my dev environment in DBeaver is:
G:\My Drive\dev_SNOWFLAKE\DEV__XRAY

Im not using DBeaver projects neither workspaces, I just drag directly files from my Google Drive into it, to edit and execute against Snowflake database.

Does above findings help? maybe some Dbeaver option already exists to workaround EFS java library problem ?

Thanks

**Comment by emanueol** [comment]
SOLUTION: Ok heres the workaround I found, it needs 3+1 things:
1. using workspaces (doesn't matter if workspace created inside or outside Google Drive) tho.
2. workspace > "Link Folder" option
3. and 2 settings needed (native hooks=OFF & on access=ON)

4. Important as well "Dont drag & drop": Any scripts must be open from dbeaver > workspace, linked folder. So do NOT drag & drop files into DBEaver, EVEN if they part of a Linked Folder!!
![image](https://github.com/user-attachments/assets/7d2c4a04-c30a-4c23-ba8d-97785ebdd957)


-- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - 
1. so imagine my workspace at:
C:\Users\Emanuel\AppData\Roaming\DBeaverData\workspace6
meaning by default new SQL Scripts are created under:
C:\Users\Emanuel\AppData\Roaming\DBeaverData\workspace6\General\Scripts

But I have my files in google drive at:
G:\My Drive\dev_SNOWFLAKE\DEV__XRAY

2. So the trick is to "link folder" on workspace to our scripts folder:
window > Projects > General > Scripts > [right click] > Create Link Folder > give path:
G:\My Drive\dev_SNOWFLAKE\DEV__XRAY

3. the two settings:
and don't forget the magic workaround option I found/shared in previous post:
window > preferences > search hook ? > workspace:
  [ ] Refresh using native hooks or polling  <--- must be off
  [x] Refresh on access                      <--- must be on


And that's it, "damm" popup is gone :) Finally!!!!

Im sure community will be very happy with this one, oh gosh took me 2 years to figure this out.
As I been only drag/drop SQL script to edit o DBeaver, never cared with Workspaces, but somehow was able to link couple of things here and there to figure out the solution :)

-- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - -- - 
Heres animated gif replicating problem and the fix:
- open dbeaver and directly drag/drop script from file explorer
- several CTRL+S throwing the "damm" popup
then fix by:
- dbeaver > workspace > link folder
- dbeaver > two settings (1 off and 1 on)
- close and drag/drop again same script and no problem.

![dbeaver gdrive fix](https://github.com/user-attachments/assets/05f3bec1-4d7b-4a14-a1de-c1b2444e69ad)


DBeaver peeps, shall I mark this issue closed or based on my findings of problems with underneath "FSE java libraries" vs "Google Drive" its something out of your scope to pursue, like asking Google to improve their windows Google Drive desktop client ? or maybe now that you know the problem, I wonder if there's any high level stuff DBeaver devs can do to prevent ?

Let me know if shall I close this issue then ?

**Comment by bomuva** [comment]
Hi! 
Has anyone found a solution to how to work with files on Google Drive so that the "update conflict" message does not appear? Maybe it would be easier to add the "do not check for file changes in the operating system" option to the settings?

**Comment by ShadelessFox** [comment]
@bomuva I believe the solution is described in the comment above https://github.com/dbeaver/dbeaver/issues/16952#issuecomment-2433680619

> Maybe it would be easier to add the "do not check for file changes in the operating system" option to the settings?

We can't override this behavior. It comes from the Eclipse platform on which DBeaver is built.

**Comment by ChemaZapiens** [comment]
This was happening with every query until I disabled `Save on query execute` on Preferences > Editors > SQL Editor.  Now I use Ctrl+S sparingly and it no longer bothers me.

Pro tip: enable `Save on query` for a week and you'll be happy with it just happening with Ctrl+S afterwards ;) (aka the Horacio Quiroga plunge: if you feel afraid on the canoe, just jump into the Amazon for a minute).
